### PR TITLE
Add universal methods

### LINKED ISSUE #215: Make API uniform
### What is the expected enhancement?
Currently, many methods require prefixing the name with the type of the graph (i.e. `digraph_`). Would it be possible to dispatch the right method based on the type of the graph automatically? This arose in a discussion on a prior PR (https://github.com/Qiskit/qiskit-terra/pull/5471#discussion_r536334895).

One pythonic way to implement this would be to switch on the type, if implementing this in Rust is a problem. Then the right Rust method may be called by the Python code.

### PR DESCRIPTION
**Pull request by mtreinish**
In the rust generated Python API we need to have fixed class inpurts to
satisfy the traits used by the pyo3 macro generated FFI functions. This
results in duplicate methods like digraph_dfs_edges and graph_dfs_edges
with the same implementation just differing input types. To simplify the
API for users this commit adds universal functions to the python side of
the retworkx package to take in any retworkx graph object and dispatch to
the proper function in the rust generated api that relies on strict
input types.

Fixes #215

- [x] Add release notes

### CODE REVIEW

File: retworkx/__init__.py [hunk 1]
```diff
@@ -81,3 +82,356 @@ class PyDAG(PyDiGraph):
...
+        output distance matrix.
+
+    :returns: The distance matrix
+    :rtype: numpy.ndarray
+    """
+    raise TypeError("Invalid Input Type %s for graph" % type(graph))
+
+
+@distance_matrix.register(PyDiGraph)
+def _digraph_distance_matrix(graph, parallel_threshold=300):
```
Comment (lcapelluto) [inline retworkx/__init__.py:116]: I'm guessing this matches the default for `digraph_distance_matrix`?

**Review by lcapelluto — APPROVED** [review]
This is so awesome!! 🎉 
The only "oof" I have is: if you have to fix a typo in a docstring, or if you want to change a default value, you have to change it now in _three_ places 😬

File: retworkx/__init__.py (hunk 1 above)
Comment (mtreinish) [reply retworkx/__init__.py:116]: Yeah: https://github.com/Qiskit/retworkx/blob/master/src/lib.rs#L1285-L1287

Comment (mtreinish) [reply retworkx/__init__.py:116]: but that does point out I'm missing `as_undirected` here. I'll fix that now

**Comment by mtreinish** [comment]
> This is so awesome!! tada
> The only "oof" I have is: if you have to fix a typo in a docstring, or if you want to change a default value, you have to change it now in _three_ places grimacing

Yeah I couldn't figure out a way around that. :( We still need to export the per type functions both so they're accessible for the dispatch functions and for backwards compat. If we're going to be exporting the functions we'll want them to be documented.

File: retworkx/__init__.py (hunk 1 above)
Comment (mtreinish) [reply retworkx/__init__.py]: Done in: https://github.com/Qiskit/retworkx/pull/241/commits/fa90541e841f0c9957ec2e6783fb0d77c1529da2

**Pull request merged**